OsmAnd/src/net/osmand/aidl/OsmandAidlApi.java

```diff
@@ -1960,54 +1967,66 @@ public boolean processResult(GPXFile result) {
 		gpxAsyncLoaderTask.executeOnExecutor(AsyncTask.THREAD_POOL_EXECUTOR);
 	}
 
-	private Map<String, FileOutputStream> copyFilesCache = new ConcurrentHashMap<>();
+	private Map<String, Map <Integer, FileOutputStream>>copyFilesCache = new ConcurrentHashMap<>();
+	private AtomicInteger counter = new AtomicInteger(0);
+
+
+	public int getNextUniqueIndex() {
+		return counter.getAndIncrement();
+	}
 
-	boolean copyFile(final CopyFileParams filePart) {
+	int copyFile(final CopyFileParams filePart) {
 		if (Algorithms.isEmpty(filePart.getFilename()) || filePart.getFilePartData() == null) {
-			return false;
+			return FILE_PARAMS_ERROR;
 		}
 
 		if (filePart.getFilename().endsWith(IndexConstants.SQLITE_EXT)) {
 			return copyFileImpl(filePart, app.getAppPath(
-				IndexConstants.TEMP_DIR + filePart.getCopyStartTime() + "/" + filePart.getFilename()),
+				IndexConstants.TEMP_DIR + filePart.getFilename()),
 				IndexConstants.TILES_INDEX_DIR);
+		} else {
+			return UNKNOWN_FILE_TYPE;
 		}
-		return false;
 	}
 
-	private boolean copyFileImpl(CopyFileParams fileParams, File file, String destination){
+	private int copyFileImpl(CopyFileParams fileParams, File file, String destination){
 
 		if (fileParams.getFilePartData().length > 256*1024) {
-			return false;
+			return LARGE_PARTS_ERROR;
 		}
 
 		FileOutputStream fos;
-		String key = fileParams.getFilename() + fileParams.getCopyStartTime();
+		String key = fileParams.getFilename();
 		try {
-			if (copyFilesCache.containsKey(key)) {
-				fos = copyFilesCache.get(key);
+			if (copyFilesCache.containsKey(key) && fileParams.getId()!=0) {
+				fos = copyFilesCache.get(key).get(fileParams.getId());
+
 				if (fileParams.isTransmitComplete()) {
 					copyFilesCache.remove(key);
 					fos.close();
 					file.renameTo(app.getAppPath(destination + fileParams.getFilename()));
-					return true;
+					return fileParams.getId();
 
 				} else {
 					fos.write(fileParams.getFilePartData());
-					return true;
+					return fileParams.getId();
 				}
-
+			} else if (copyFilesCache.containsKey(key) && fileParams.getId()==0){
+				return DOUBLE_COPY_ERROR;
 			} else {
 				file.getParentFile().mkdirs();
 				fos = new FileOutputStream(file, true);
 				fos.write(fileParams.getFilePartData());
-				copyFilesCache.put(key, fos);
-				return true;
+				counter.getAndIncrement();
```

Comment:
no variable assignment ? the code below 
```
counter.intValue();
```
is incorrect because of race condition